crates/claims/crates/jws/src/compact/str.rs

```diff
@@ -1,29 +1,89 @@
+use base64::Engine;
 use core::fmt;
 use std::{ops::Deref, str::FromStr};
 
-use base64::Engine;
-
-use crate::{CompactJWS, DecodeError, DecodedJWS, DecodedSigningBytes, Header, InvalidCompactJWS};
+use crate::{
+    utils::is_url_safe_base64_char, DecodeError, DecodedJws, Header, InvalidJws, JwsSlice,
+};
 
-/// JWS in UTF-8 compact serialized form.
+/// Borrowed UTF-8 encoded JWS.
+///
+/// This is an unsized type borrowing the JWS and meant to be referenced as
+/// `&JwsStr`, just like `&str`.
+/// Use [`JwsString`] if you need to own the JWS.
 ///
-/// Contrarily to [`CompactJWS`], this type guarantees that the payload is
-/// a valid UTF-8 string, meaning the whole compact JWS is an UTF-8 string.
-/// This does not necessarily mean the payload is base64 encoded.
+/// This type is similar to the [`Jws`](crate::Jws) type.
+/// However contrarily to `Jws`, there is no guarantee that the JWS is URL-safe.
+///
+/// Use [`Jws`](crate::Jws) if you expect URL-safe JWSs.
+#[derive(PartialEq, Eq, PartialOrd, Ord, Hash)]
 #[repr(transparent)]
-pub struct CompactJWSStr(CompactJWS);
-
-impl CompactJWSStr {
-    pub fn new(data: &[u8]) -> Result<&Self, InvalidCompactJWS<&[u8]>> {
-        match std::str::from_utf8(data) {
-            Ok(s) => Self::from_string(s).map_err(|_| InvalidCompactJWS(data)),
-            Err(_) => Err(InvalidCompactJWS(data)),
+pub struct JwsStr(JwsSlice);
+
+impl JwsStr {
+    pub fn new<T: ?Sized + AsRef<[u8]>>(data: &T) -> Result<&Self, InvalidJws<&T>> {
+        let bytes = data.as_ref();
+        match std::str::from_utf8(bytes) {
+            Ok(_) => {
+                let _ = JwsSlice::new(bytes).map_err(|_| InvalidJws(data))?;
+                Ok(unsafe { Self::new_unchecked(bytes) })
+            }
+            Err(_) => Err(InvalidJws(data)),
         }
     }
 
-    pub fn from_string(data: &str) -> Result<&Self, InvalidCompactJWS<&str>> {
-        let inner = CompactJWS::new(data.as_bytes()).map_err(|_| InvalidCompactJWS(data))?;
-        Ok(unsafe { std::mem::transmute::<&CompactJWS, &Self>(inner) })
+    pub const fn validate(bytes: &[u8]) -> bool {
+        Self::validate_range(bytes, 0, bytes.len())
+    }
+
+    pub const fn validate_range(bytes: &[u8], mut i: usize, end: usize) -> bool {
```

Comment:
Looking into this, I think the `const` method `SdJwt::validate` needs access to `Jws::validate_range`. And since we can take a sliced range from the bytes input, we can't use `Jws::validate`.


Please disregard the question, unless you feel different. I think it is fine as is.

Reply:
Yes I introduced `validate_range` to use in `SdJwt::validate` because we can't borrow a subslice from a range in `const` context.